kaska/s2_observations.py

```diff
@@ -219,133 +263,181 @@ def read_time_series(self, time_grid):
         return s2_obs
 
     def read_granule(self, timestep):
-        """Reads data granule for a given `timestep`. Returns all relevant
-        bits and bobs (surface reflectrance, angles, cloud mask, uncertainty).
+        """Reads the data for the given timestep. Returns all relevant
+        quantities (surface reflectrance, angles, cloud mask, uncertainty).
         The mask is true for OK pixels. If there are no suitable pixels, the
         returned tuple is a collection of `None`
 
-
         Parameters
         ----------
-        timestep : datetime
-            The datetime object
-
+        timestep : A datetime object.
         Returns
         -------
-        tuple
-            rho_surface, mask, sza, vza, raa, rho_unc
-        """
+        tuple containing the rho_surface, mask, sza, vza, raa, rho_unc
 
+        NOTE: Currently reads in sequentially. It's better to gather
+        all the filenames and read them in parallel using parmap.py
+        """
         assert timestep in self.date_data, f"{str(timestep):s} not available!"
-        # NOTE: Currently reads in sequentially. It's better to gather
-        # all the filenames and read them in parallel using parmap.py"""
-        current_folder = self.date_data[timestep]
-        fname_prefix = [
-            f.name.split("B02")[0] for f in current_folder.glob("*B02_sur.tif")
-        ][0]
-        # Find cloud mask
-        cloud_mask = current_folder.parent / f"cloud.tif"
+
+        # This is the parent folder for this granule
+        current_folder = Path()
+        for part in self.date_data[timestep][0].parts:
+            current_folder = current_folder / part
+            if ".SAFE" in part:
+                break
+
+        # Read in cloud mask and reproject it on state mask.
+        # The cloud mask is the probabilty of cloud. OK pixels have
+        # a probability of cloud below `band_prob_threshold`.
+        # Stop processing if cloud mask file doesn't exist,
+        # or if no clear pixels exist.
+        try:
+            cloud_mask = next(current_folder.glob("**/cloud.tif"))
+        except StopIteration:
+            LOG.info(f"Cloud file cloud.tif does not exist.")
+            #print(f"Cloud file cloud.tif does not exist.")
+            return None, None, None, None, None, None
         cloud_mask = reproject_data(
-            str(cloud_mask), target_img=self.state_mask
-        ).ReadAsArray()
-        # cloud mask is probabilty of cloud
-        # OK pixels have a probability of cloud below `band_prob_threshold`
+            str(cloud_mask), target_img=self.state_mask).ReadAsArray()
```

Comment:
I don't want to change the arguments of `reproject_data`, because it's called by `s1_observations` as well, with a funky-looking string as a file argument (see line 119 or so) - I've no idea of how that corresponds to a path.
I'll take care of the other 2 comments.